contractsV5/KyberHintHandler.sol

```diff
@@ -216,84 +247,103 @@ contract KyberHintParser is IKyberHint, Utils {
-            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth);
-        } else if (opcodeHash == keccak256(parseOpcode(TradeType.Split))) {
-            reserves.tradeType = TradeType.Split;
-            (reserves.addresses, reserves.splitValuesBps, indexToContinueFrom) = decodeReservesFromHint(true, hint, indexToContinueFrom);
-            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth);
-        } else {
-            revert("Invalid hint opcode");
+            decodeOperation(hint, tradeHint, indexToContinueFrom, false, isExternal);
+        } else if (opcodeHash == keccak256(encodeOpcode(TradeType.MaskIn))) {
+            reservesHint.tradeType = TradeType.MaskIn;
+            (indexToContinueFrom) = decodeReservesFromHint(false, hint, reservesHint, isExternal, indexToContinueFrom);
+            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth, isExternal);
+        } else if (opcodeHash == keccak256(encodeOpcode(TradeType.MaskOut))) {
+            reservesHint.tradeType = TradeType.MaskOut;
+            (indexToContinueFrom) = decodeReservesFromHint(false, hint, reservesHint, isExternal, indexToContinueFrom);
+            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth, isExternal);
+        } else if (opcodeHash == keccak256(encodeOpcode(TradeType.Split))) {
+            reservesHint.tradeType = TradeType.Split;
+            (indexToContinueFrom) = decodeReservesFromHint(true, hint, reservesHint, isExternal, indexToContinueFrom);
+            decodeOperation(hint, tradeHint, indexToContinueFrom, isTokenToEth, isExternal);
+        } else {            
+            reservesHint.invalidHint = true;
         }
     }
 
     function decodeReservesFromHint(
         bool isTokenSplit,
         bytes memory hint,
+        ReservesHint memory reservesHint,
+        bool isExternal,
         uint indexToContinueFrom
     )
         internal
         view
         returns (
-            IKyberReserve[] memory,
-            uint[] memory,
             uint
         )
     {
        uint reservesLength = hint.toUint8(indexToContinueFrom);
        uint bpsSoFar;
        uint[] memory splitValuesBps;
-
        IKyberReserve[] memory addresses = new IKyberReserve[](reservesLength);
+       bytes8[] memory reserveIds = new bytes8[](reservesLength);
+
        if (isTokenSplit) {
            splitValuesBps = new uint[](reservesLength);
        } else {
            splitValuesBps = new uint[](1);
            splitValuesBps[0] = BPS;
        }
+
        indexToContinueFrom++;
 
-       for (uint i = 0; i < reservesLength; i++) {           
-           addresses[i] = IKyberReserve(convertReserveIdToAddress(hint.slice(indexToContinueFrom, RESERVE_ID_LENGTH).toBytes8(0)));
+       for (uint i = 0; i < reservesLength; i++) {
+           reserveIds[i] = hint.slice(indexToContinueFrom, RESERVE_ID_LENGTH).toBytes8(0);
+
+           if (!isExternal) {
```

Comment:
since parsing external is running off chain.
this code shouldn't be here
it should be on the external call
we don't care to have another loop that translates all addresses to IDs.

Reply:
Implemented.